### PR TITLE
feat: generate a templated Numaflow controller definition manifest based on the latest version of the Numaflow numaspace-install manifest

### LINKED ISSUE #373: Add capability to easily add a new templated Numaflow Controller manifest to be deployed when running "make start"
# Summary

Add capability to easily add a new templated Numaflow Controller manifest to be deployed when running "make start"

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by afugazzotto**
Fixes #373 

### Modifications

Created a script that uses kustomize and yq to generate a templated Numaflow controller definition manifest based on the latest version of the Numaflow numaspace-install manifest. The script can be called with an argument to generate a specified file or without an argument to generate the yaml file in the default controller definitions manifest directory (generated via `make start`).

### Verification

Manual tests performed on the script (both with and without argument) and tested the generated yaml in a cluster to validate the generated configmap.

### CODE REVIEW

File: tests/manifests/default/controller_def_1.3.3.yaml [hunk 1]
```diff
@@ -1,325 +1,332 @@
 apiVersion: v1
```
Comment (afugazzotto) [inline tests/manifests/default/controller_def_1.3.3.yaml:1]: @juliev0 do we also want to autogenerate the `-copy1` file equivalent?

Comment (juliev0) [reply tests/manifests/default/controller_def_1.3.3.yaml:1]: no, that's okay, thank you!

File: hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+BASE_DIR=../..
+TEST_MANIFEST_DIR=tests/manifests/default
+SCRIPT_DIR=$(dirname "$0")
+
+# Change working directory
+cd $SCRIPT_DIR
+
+# Get the latest Numaflow Version from the remote repository
+echo "Getting the latest Numaflow version tag..."
+NUMAFLOW_VERSION=$(git ls-remote --tags https://github.com/numaproj/numaflow | grep -v '{}' | tail -1 | sed 's/.*\///' | sed 's/^v//')
```
Comment (juliev0) [inline hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh:17]: Ideally we should be able to run this script for any Git tag and not just the latest, right? (but I'm fine if done in a follow up)

Comment (afugazzotto) [reply hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh:17]: I'll add an argument to use a specific tag if present or to use latest if not present.

File: hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+# Install yq if not present
+if ! command -v yq 2>&1 >/dev/null
+then
+  echo "yq not found, installing it..."
+  go install github.com/mikefarah/yq/v4@latest
+  echo "yq installed"
+fi
+
+# Use yq to modify/add the instance field of the string litteral yaml config of the numaflow-controller-config ConfigMap data field controller-config.yaml
+# This is not possible via Kustomize yet. Follow https://github.com/kubernetes-sigs/kustomize/issues/4517 and https://github.com/kubernetes-sigs/kustomize/pull/5679 for future Kustomize updates.
```
Comment (juliev0) [inline hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh:53]: thanks for finding a reference for that!

File: hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+# Install yq if not present
+if ! command -v yq 2>&1 >/dev/null
+then
+  echo "yq not found, installing it..."
+  go install github.com/mikefarah/yq/v4@latest
+  echo "yq installed"
+fi
+
+# Use yq to modify/add the instance field of the string litteral yaml config of the numaflow-controller-config ConfigMap data field controller-config.yaml
```
Comment (juliev0) [inline hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh:52]:
```suggestion
# Use yq to modify/add the instance field of the string literal yaml config of the numaflow-controller-config ConfigMap data field controller-config.yaml
```

File: hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+yq 'select(.kind == "ConfigMap" and .metadata.name == "numaflow-controller-config{{ .InstanceSuffix }}") |= .data."controller-config.yaml" = strenv(TMP_NFC_DEF_GEN)' tmp-kustomized-nfinstall.yaml > tmp-full-spec.yaml
+
+# Update the version on the base definitions configmap yaml
+export TMP_NFC_DEF_GEN_NUMAFLOW_VERSION=$NUMAFLOW_VERSION
+envsubst < def-configmap.yaml > tmp-versioned-def-cm.yaml
+
+# Using yq, append tmp-full-spec.yaml to tmp-versioned-def-cm.yaml as a string literal under the controllerDefinitions fullSpec field for the specified version
+export TMP_NFC_DEF_GEN_FULL_SPEC=$(cat tmp-full-spec.yaml | yq)
+export TMP_NFC_DEF_GEN_FULL_SPEC_PLUS=$(yq '.data."controller_definitions.yaml" | fromyaml | .controllerDefinitions[0].fullSpec = strenv(TMP_NFC_DEF_GEN_FULL_SPEC)' tmp-versioned-def-cm.yaml)
+yq '.data."controller_definitions.yaml" = strenv(TMP_NFC_DEF_GEN_FULL_SPEC_PLUS)' tmp-versioned-def-cm.yaml > $BASE_DIR/$OUTPUT_FILE
```
Comment (juliev0) [inline hack/numaflow-controller-def-generator/numaflow-controller-def-generator.sh:64]: `yq` seems like a great idea - nice job getting that to work

File: hack/numaflow-controller-def-generator/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+commonLabels:
+  app.kubernetes.io/instance: '{{ .InstanceID }}'
+nameSuffix: "{{ .InstanceSuffix }}"
```
Comment (juliev0) [inline hack/numaflow-controller-def-generator/kustomization.yaml:5]: wow, so other than the `instance` field in the `numaflow-controller-config`, these lines took care of everything - nice job!

Comment (afugazzotto) [reply hack/numaflow-controller-def-generator/kustomization.yaml:5]: thanks! yes, those were some nice findings

**Review by juliev0 — APPROVED** [review]

**Pull request merged**